### ISSUE TITLE
Potential bug in HybridPoseEstimator?

### ISSUE THREAD

**Issue by jtressle** [PoseLib/PoseLib#117]
Hi,

First off, great library.  It's executed very well.

I had a question regarding the sampling in HybridPoseEstimator?

In the header, num_data is defined as:

```
   HybridPoseEstimator(const RansacOptions &ransac_opt, const std::vector<Point2D> &points2D,
                        const std::vector<Point3D> &points3D, const std::vector<PairwiseMatches> &pairwise_matches,
                        const std::vector<CameraPose> &map_ext)
        : opt(ransac_opt), x(points2D), X(points3D), matches(pairwise_matches), map_poses(map_ext) {
        rng = opt.seed;
        xs.resize(sample_sz);
        Xs.resize(sample_sz);
        sample.resize(sample_sz);
        num_data = points2D.size();
        for (const PairwiseMatches &m : matches) {
            num_data += m.x1.size();
        }
    }
```
    
However, the sampler is only collecting points from `x` and `X` here:

```
void HybridPoseEstimator::generate_models(std::vector<CameraPose> *models) {
    draw_sample(sample_sz, num_data, &sample, rng);
    for (size_t k = 0; k < sample_sz; ++k) {
        xs[k] = x[sample[k]].homogeneous().normalized();
        Xs[k] = X[sample[k]];
    }
    p3p(xs, Xs, models);
    // TODO: actual hybrid sampling (we have p2p2pl and 5+1 gen-relpose already implemented, should be enough)
}
```

Shouldn't `num_data` only be set to points2D.size()?

Thanks

**Comment by jtressle** [comment]
FYI, I just tested changing num_data to just be points2D.size(), and there was minuscule differences in results.  Basically about 0.5% difference in inliers.  So it's not a critical issue.

**Comment by pablospe** [comment]
Thanks! Could you propose a PR with your suggested changes? That would help.

**Comment by vlarsson** [comment]
Fixed in #126.

**Comment by jtressle** [comment]
Hi @vlarsson,

Sorry for not doing this earlier, and thanks for issuing the fix.  

I took a look at the code and I think this might be a better fix, where I just updated draw_sample to use up to x.size() points:

`void HybridPoseEstimator::generate_models(std::vector<CameraPose> *models) {
    draw_sample(sample_sz, x.size(), &sample, rng);
    for (size_t k = 0; k < sample_sz; ++k) {
        xs[k] = x[sample[k]].homogeneous().normalized();
        Xs[k] = X[sample[k]];
    }
    p3p(xs, Xs, models);
    // TODO: actual hybrid sampling (we have p2p2pl and 5+1 gen-relpose already implemented, should be enough)
}`

The reason is in the ransac_impl.h file, the num_data is used here:
`stats.inlier_ratio = static_cast<double>(stats.num_inliers) / static_cast<double>(estimator.num_data);`

Since the HybridPoseEstimator `score_model` adds up both the point2D-point3D inliers and all the matching inliers, I think the num_data is correctly computed.  Sorry, I didn't see this issue when I first described the problem.

Does this make sense? 

Thanks,
Jan-Michael
